_pytest/config.py

```diff
@@ -1257,9 +1257,13 @@ def get_common_ancestor(args):
 
 
 def get_dirs_from_args(args):
-    return [d for d in (py.path.local(x) for x in args
-                        if not str(x).startswith("-"))
-            if d.exists()]
+    possible_dirs = [
+        py.path.local(str(arg).split('::')[0])
+        for arg in args
+        if not str(arg).startswith('-')
+    ]
+
+    return [d for d in possible_dirs if d.exists()]
```

Comment:
Why don't you like my generator expression?! ;)
In your case `str` is still called twice on `arg`.. but I can see that readability is better with a named var here.
Nitpick: remove the added newline to keep it more compact.